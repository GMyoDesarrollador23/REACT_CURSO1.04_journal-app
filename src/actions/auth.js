import Swal from "sweetalert2";

import { types } from "../types/types";
import {
   firebase,
   googleAuthProvider,
} from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";

import { notesLogout } from "../actions/notes";

export const startLoginEmailPassword = (
   email,
   password
) => {
   return (dispatch) => {
      dispatch(startLoading());
      return firebase
         .auth()
         .signInWithEmailAndPassword(email, password)
         .then(({ user }) => {
            dispatch(login(user.uid, user.displayName));

            dispatch(finishLoading());
            Swal.fire({
               position: "center",
               icon: "success",
               title: `welcome back "${user.displayName}"`,
               showConfirmButton: false,
               timer: 2000,
            });
         })
         .catch((e) => {
            console.warn(e);
            dispatch(finishLoading());
            Swal.fire({
               position: "center",
               icon: "error",
               title: e.message,
               showConfirmButton: false,
               timer: 2000,
            });
         });
   };
};

export const startRegisterWithEmailPasswordName = (
   email,
   password,
   name
) => {
   return (dispatch) => {
      firebase
         .auth()
         .createUserWithEmailAndPassword(email, password)
         .then(async ({ user }) => {
            await user.updateProfile({
               displayName: name,
            });
            // console.log(user);

            dispatch(login(user.uid, user.displayName));
         })
         .catch((e) => {
            console.log(e);
            Swal.fire({
               position: "center",
               icon: "error",
               title: e.message,
               showConfirmButton: false,
               timer: 2000,
            });
         });
   };
};

export const startGoogleLogin = () => {
   return (dispatch) => {
      firebase
         .auth()
         .signInWithPopup(googleAuthProvider)
         .then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
         });
   };
};

export const startLogoud = () => {
   return async (dispatch) => {
      await firebase.auth().signOut();

      dispatch(logout());
      dispatch(notesLogout());
   };
};

export const login = (uid, displayName) => ({
   type: types.login,
   payload: {
      uid,
      displayName,
   },
});

export const logout = () => ({
   type: types.logout,
});
