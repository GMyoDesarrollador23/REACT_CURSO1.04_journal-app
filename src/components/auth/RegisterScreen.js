import React from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import validator from "validator";

import { useForm } from "../../hooks/useForm";
import { setError, removeError } from "../../actions/ui";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";

export const RegisterScreen = () => {
   const dispatch = useDispatch();

   const { msgError } = useSelector((state) => state.ui);

   const [FormValues, handleInputChange] = useForm({
      name: "Gerardo",
      email: "ddaalgo@algo.sds",
      password: "123456",
      password2: "123456",
   });

   const { name, email, password, password2 } = FormValues;

   const handleRegister = (e) => {
      e.preventDefault();
      if (isFromValid()) {
         dispatch(
            startRegisterWithEmailPasswordName(
               email,
               password,
               name
            )
         );
      }
   };

   const isFromValid = () => {
      if (name.trim().length === 0) {
         dispatch(setError("name is required"));
         return false;
      } else if (!validator.isEmail(email)) {
         dispatch(setError("email is not valid"));
         return false;
      } else if (
         password !== password2 ||
         password.length < 5
      ) {
         dispatch(
            setError(
               "Password should be at least 6 characters and match each other"
            )
         );
         return false;
      }

      dispatch(removeError());
      return true;
   };

   return (
      <>
         <h3 className="auth__title animate__animated animate__fadeIn animate__faster">
            Register
         </h3>
         <form
            onSubmit={handleRegister}
            className="animate__animated animate__fadeIn animate__faster"
         >
            {msgError && (
               <div className="auth__alert-error">
                  {msgError}
               </div>
            )}
            <input
               type="text"
               placeholder="Name"
               autoComplete="off"
               name="name"
               className="auth__input"
               onChange={handleInputChange}
               value={name}
            />

            <input
               type="email"
               placeholder="Email"
               name="email"
               autoComplete="off"
               className="auth__input"
               onChange={handleInputChange}
               value={email}
            />

            <input
               type="password"
               placeholder="Password"
               name="password"
               autoComplete="off"
               className="auth__input"
               onChange={handleInputChange}
               value={password}
            />

            <input
               type="password"
               placeholder="Confim Password"
               name="password2"
               className="auth__input"
               onChange={handleInputChange}
               value={password2}
            />

            <button
               type="submit"
               className="btn btn-primary btn-block mb-5"
               // disabled={true}
            >
               Register
            </button>
            <Link to="/auth/login" className="links">
               Already Register?
            </Link>
         </form>
      </>
   );
};
