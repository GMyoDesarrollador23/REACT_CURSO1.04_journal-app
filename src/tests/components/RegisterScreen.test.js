import React from "react";
import { mount } from "enzyme";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

import { RegisterScreen } from "../../components/auth/RegisterScreen";

// *****************************************************************
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
   auth: {},
   ui: {
      loading: false,
      msgError: null,
   },
};

let store = mockStore(initState);

const wraper = mount(
   <Provider store={store}>
      <MemoryRouter>
         <RegisterScreen />
      </MemoryRouter>
   </Provider>
);
// *****************************************************************
describe("Pruebas en el <RegisterScreen/>", () => {
   // ----------------------------------------------------------

   beforeEach(() => {
      store = mockStore(initState);
      jest.clearAllMocks();
   });

   // ----------------------------------------------------------
   test("debe de mostrarse correctamente", () => {
      expect(wraper).toMatchSnapshot();
   });

   // ----------------------------------------------------------
   test("debe de hacer el dipatch de la accion respectiva", () => {
      const emailField = wraper.find('input[name="email"]');

      emailField.simulate("change", {
         target: {
            value: "",
            name: "email",
         },
      });

      wraper.find("form").simulate("submit", {
         preventDefault() {},
      });
      const actions = store.getActions();
      expect(actions).toEqual([]);
   });

   // ----------------------------------------------------------
   test("debe de mostrar la caja de alerta con el error", () => {
      const initState = {
         auth: {},
         ui: {
            loading: false,
            msgError: "email no es correcto",
         },
      };

      const store = mockStore(initState);

      const wraper = mount(
         <Provider store={store}>
            <MemoryRouter>
               <RegisterScreen />
            </MemoryRouter>
         </Provider>
      );

      expect(
         wraper.find(".auth__alert-error").exists()
      ).toBe(true);

      expect(
         wraper.find(".auth__alert-error").text().trim()
      ).toBe(initState.ui.msgError);
   });
   // ----------------------------------------------------------
});
