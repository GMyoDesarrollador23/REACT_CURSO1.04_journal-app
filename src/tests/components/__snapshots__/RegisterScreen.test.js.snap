// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <RegisterScreen/> debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title animate__animated animate__fadeIn animate__faster"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Gerardo"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value="ddaalgo@algo.sds"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confim Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="links"
      href="/auth/login"
      onClick={[Function]}
    >
      Already Register?
    </a>
  </form>,
]
`;
